src/vivarium_public_health/metrics/utilities.py

```diff
@@ -148,6 +149,85 @@ def get_output_template(by_age: bool, by_sex: bool, by_year: bool) -> OutputTemp
     return OutputTemplate(template)
 
 
+def get_age_sex_filter_and_iterables(config: dict, age_bins: pd.DataFrame) -> (
+        QueryString, Tuple[Iterable[Tuple[str, pd.Series]], List[str]]):
+    """Constructs a filter and a set of iterables for age and sex.
+
+    The constructed filter and iterables are based on configuration for the
+    observer component.
+
+    Parameters
+    ----------
+    config
+        A mapping with 'by_age' and 'by_sex' keys and boolean values
+        indicating whether the observer is binning data by the respective
+        categories.
+    age_bins
+        A table containing bin names and bin edges.
+
+    Returns
+    -------
+    age_sex_filter
+        A filter on age and sex for use with DataFrame.query
+    (ages, sexes)
+        Iterables for the age and sex groups partially defining the bins
+        for the observers.
+
+    """
+    age_sex_filter = QueryString("")
+    if config['by_age']:
+        ages = age_bins.set_index('age_group_name').iterrows()
+        age_sex_filter += '{age_group_start} <= age and age < {age_group_end}'
+    else:
+        ages = [('all_ages', pd.Series({'age_group_start': None, 'age_group_end': None}))]
+
+    if config['by_sex']:
+        sexes = ['Male', 'Female']
+        age_sex_filter += 'sex == "{sex}"'
+    else:
+        sexes = ['Both']
+
+    return age_sex_filter, (ages, sexes)
+
+
+def get_time_span_filter_and_iterable(config: dict, sim_start: pd.Timestamp, sim_end: pd.Timestamp) -> (
+        QueryString, List[Tuple[str, Tuple[pd.Timestamp, pd.Timestamp]]]):
+    """Constructs a filter and iterable for time.
+
+    The constructed filter and iterable are based on configuration for the
+    observer component.
+
+    Parameters
+    ----------
+    config
+        A mapping with 'by_yer' and a boolean value indicating whether
+        the observer is binning data by year.
+    sim_start
+        The time the simulation starts.
+    sim_end
+        The time the simulation ends.
+
+    Returns
+    -------
+    span_filter
+        A filter on time for use with DataFrame.query
+    time_spans
+        Iterable for the age and sex groups partially defining the bins
+        for the observers.
+
+    """
+    if config['by_year']:
+        time_spans = [(year, (pd.Timestamp(year=year, month=1, day=1), pd.Timestamp(f'1-1-{year + 1}')))
+                      for year in range(sim_start.year, sim_end.year + 1)]
+    else:
+        time_spans = [('all_years', (pd.Timestamp(f'1-1-1900'), pd.Timestamp(f'1-1-2100')))]
+    # This filter needs to be applied separately to compute additional
+    # attributes in the person time calculation.
+    span_filter = QueryString('{t_start} <= exit_time and entrance_time < {t_end}')
```

Comment:
I think I'm still a little fuzzy on how the year binning works. Why is this <= exit_time instead of < exit_time? do we assume exit_time is the end of the day or something?